src/mean.ts

```diff
@@ -0,0 +1,87 @@
+import type { IterableContainer } from "./internal/types";
+import { purry } from "./purry";
+import { sum } from "./sum";
+
+type Mean<T extends IterableContainer<bigint> | IterableContainer<number>> =
+  // Empty arrays would always result in undefined.
+  T extends readonly []
+    ? undefined
+    : // Non-empty bigint arrays will always result in bigint.
+      T extends readonly [bigint, ...ReadonlyArray<unknown>]
+      ? bigint
+      : // But an empty bigint array would result in undefined.
+        T[number] extends bigint
+        ? undefined
+        : // Non-bigint arrays are always handled correctly.
+          number;
```

Comment:
**question:** shouldn't this be `number | undefined`?

Reply:
You are right, thanks, I missed that. I am going to simplify this type, taking into account the rest of your suggestions.